File: src/main/scala/aoc/Day02.scala
package aoc

import aoc.Day02.Outcome.{ Lose, Draw, Win }
import aoc.Day02.{ Outcome, Strategy }
import aoc.Day02.Strategy.{ Rock, Paper, Scissors }

object Day02:
  type Guide = List[EncryptedRound]
  object Guide:
    def parse(lines: Iterator[String]): Guide = lines.map(EncryptedRound.parse).toList

  val StrategyCode: Map[Char, Strategy] = Map(
    'A' -> Rock,
    'B' -> Paper,
    'C' -> Scissors,
    'X' -> Rock,
    'Y' -> Paper,
    'Z' -> Scissors,
  )
  val OutcomeCode: Map[Char, Outcome] = Map(
    'X' -> Lose,
    'Y' -> Draw,
    'Z' -> Win,
  )

  case class EncryptedRound(opponent: Char, other: Char):
    def decrypt1: Round1 = Round1(StrategyCode(opponent), StrategyCode(other))
    def decrypt2: Round2 = Round2(StrategyCode(opponent), OutcomeCode(other))

  object EncryptedRound:
    def parse(s: String): EncryptedRound = s.split(" ") match
      case Array(opponent, yours) => EncryptedRound(opponent.nn.head, yours.nn.head)
      case _ => ???

  case class Round1(opponentStrategy: Strategy, myStrategy: Strategy):
    def score: Int =
      val outcome = myStrategy.playAgainst(opponentStrategy)
      outcome.points + myStrategy.points

  case class Round2(opponentStrategy: Strategy, outcome: Outcome):
    def score: Int =
      val myStrategy = opponentStrategy.strategyProducing(outcome)
      myStrategy.points + outcome.points

  enum Outcome:
    case Lose, Draw, Win
    def points: Int = ordinal * 3

  enum Strategy:
    case Rock, Paper, Scissors
    def points: Int = ordinal + 1

    def playAgainst(other: Strategy): Outcome = (this, other) match
      case (left, right) if left == right => Outcome.Draw
      case (Rock, Scissors) | (Paper, Rock) | (Scissors, Paper) => Outcome.Win
      case _ => Outcome.Lose

    def strategyProducing(outcome: Outcome): Strategy =
      Strategy.values.find(_.playAgainst(this) == outcome).get

  def part1(guide: Guide): Int = guide.map(_.decrypt1.score).sum

  def part2(guide: Guide): Int = guide.map(_.decrypt2.score).sum

  def main(args: Array[String]): Unit =
    val input = Guide.parse(Inputs.linesIterator(day = 2))
    println(part1(input))
    println(part2(input))

File: rust/src/day02.rs
use std::fs::File;
use std::io;
use std::io::BufRead;
use std::path::Path;

fn read_input() -> Vec<(char, char)> {
    let mut input = Vec::new();
    let path = Path::new("inputs/day02.txt");
    let file = File::open(path).expect("cannot open file");
    let reader = io::BufReader::new(file);
    for line in reader.lines() {
        let line = line.expect("cannot read line");
        input.push((line.chars().nth(0).unwrap(), line.chars().nth(2).unwrap()));
    }
    input
}

#[derive(PartialEq, Debug, Clone)]
enum Outcome {
    Lose,
    Draw,
    Win,
}

impl Outcome {
    fn score(&self) -> i32 {
        match *self {
            Outcome::Lose => 0,
            Outcome::Draw => 3,
            Outcome::Win => 6,
        }
    }

    fn parse(c: char) -> Outcome {
        match c {
            'X' => Outcome::Lose,
            'Y' => Outcome::Draw,
            'Z' => Outcome::Win,
            _ => panic!("cannot parse {c} into an outcome"),
        }
    }

    fn strategy_to_get(&self, their_strategy: Strategy) -> Strategy {
        match (self, their_strategy) {
            (Outcome::Draw, _) => their_strategy,
            (Outcome::Win, Strategy::Rock) => Strategy::Paper,
            (Outcome::Win, Strategy::Paper) => Strategy::Scissors,
            (Outcome::Win, Strategy::Scissors) => Strategy::Rock,
            (Outcome::Lose, Strategy::Rock) => Strategy::Scissors,
            (Outcome::Lose, Strategy::Paper) => Strategy::Rock,
            (Outcome::Lose, Strategy::Scissors) => Strategy::Paper,
        }
    }
}

#[derive(PartialEq, Debug, Clone, Copy)]
enum Strategy {
    Rock,
    Paper,
    Scissors,
}

impl Strategy {
    fn score(&self) -> i32 {
        match *self {
            Strategy::Rock => 1,
            Strategy::Paper => 2,
            Strategy::Scissors => 3,
        }
    }

    fn against(&self, other: Strategy) -> Outcome {
        match (*self, other) {
            (Strategy::Rock, Strategy::Scissors)
            | (Strategy::Paper, Strategy::Rock)
            | (Strategy::Scissors, Strategy::Paper) => Outcome::Win,
            (left, right) if left == right => Outcome::Draw,
            _ => Outcome::Lose,
        }
    }

    fn parse(c: char) -> Strategy {
        match c {
            'A' | 'X' => Strategy::Rock,
            'B' | 'Y' => Strategy::Paper,
            'C' | 'Z' => Strategy::Scissors,
            _ => panic!("cannot parse {c} into a strategy"),
        }
    }
}

fn part1(input: &Vec<(char, char)>) -> i32 {
    let mut total_score = 0;
    for tuple in input {
        let their_strategy = Strategy::parse(tuple.0);
        let my_strategy = Strategy::parse(tuple.1);
        let outcome = my_strategy.against(their_strategy);
        let score = my_strategy.score() + outcome.score();
        total_score += score
    }
    total_score
}

fn part2(input: &Vec<(char, char)>) -> i32 {
    let mut total_score = 0;
    for tuple in input {
        let their_strategy = Strategy::parse(tuple.0);
        let outcome = Outcome::parse(tuple.1);
        let my_strategy = outcome.strategy_to_get(their_strategy);
        let score = my_strategy.score() + outcome.score();
        total_score += score
    }
    total_score
}

pub fn solve() {
    let input = read_input();
    println!(
        "DAY 02\n======\npart 1: {}\npart 2: {}\n",
        part1(&input),
        part2(&input)
    );
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_strategy() {
        assert_eq!(Strategy::Rock.against(Strategy::Rock), Outcome::Draw);
        assert_eq!(Strategy::Rock.against(Strategy::Paper), Outcome::Lose);
        assert_eq!(Strategy::Rock.against(Strategy::Scissors), Outcome::Win);
        assert_eq!(Strategy::Paper.against(Strategy::Rock), Outcome::Win);
        assert_eq!(Strategy::Paper.against(Strategy::Paper), Outcome::Draw);
        assert_eq!(Strategy::Paper.against(Strategy::Scissors), Outcome::Lose);
        assert_eq!(Strategy::Scissors.against(Strategy::Rock), Outcome::Lose);
        assert_eq!(Strategy::Scissors.against(Strategy::Paper), Outcome::Win);
        assert_eq!(Strategy::Scissors.against(Strategy::Scissors), Outcome::Draw);
    }

    #[test]
    fn test_part1() {
        let input = vec![
            ('A', 'Y'),
            ('B', 'X'),
            ('C', 'Z')
        ];
        assert_eq!(part1(&input), 15);
    }

    #[test]
    fn test_part2() {
        let input = vec![
            ('A', 'Y'),
            ('B', 'X'),
            ('C', 'Z')
        ];
        assert_eq!(part2(&input), 12);
    }
}